Reemit.Common.UnitTests/SharedReaderTests.cs

```diff
@@ -14,10 +18,200 @@ public void ReadBytes_Called_AdvancesRelativeOffsetAndOffset()
         const int readBytesCount = 2;
         
         // Act
-        sharedReader.ReadBytes(readBytesCount);
+        var actualBytes = sharedReader.ReadBytes(readBytesCount);
         
         // Assert
         Assert.Equal(sharedReaderOffset + readBytesCount, sharedReader.Offset);
         Assert.Equal(readBytesCount, sharedReader.RelativeOffset);
+        Assert.Equal(readBytesCount, actualBytes.Length);
+        Assert.NotNull(actualBytes);
+        Assert.Equal([0x1, 0x2], actualBytes);
+    }
+
+    [Fact]
+    public void ReadBytes_Called_EmptyBufferKeepsRelativeOffsetAndOffset ()
+    {
+        // Arrange
+        byte[] bytes = [];
+        using var stream = new MemoryStream(bytes);
+        using var binaryReader = new BinaryReader(stream);
+        const int sharedReaderOffset = 1;
+        using var sharedReader = new SharedReader(sharedReaderOffset, binaryReader, new object());
+        const int readBytesCount = 0x100;
+
+        // Act
+        var actualBytes = sharedReader.ReadBytes(readBytesCount);
+
+        // Assert
+        Assert.Equal(sharedReaderOffset, sharedReader.Offset);
+        Assert.Equal(0, sharedReader.RelativeOffset);
+        Assert.Equal([], actualBytes);
+    }
+
+    [Theory]
+    [MemberData(nameof(GetReadUnmanagedData))]
+    public void ReadUnmanaged_Called_AdvancesRelativeOffsetAndOffset<T>(
+        int sharedReaderOffset,
+        T[] expectedUnmanagedValues,
+        Func<SharedReader, T> readUnmanaged)
+        where T : unmanaged
+    {
+        // Arrange
+        var getBytes =
+            typeof(T) == typeof(byte) ?
+                x => [Unsafe.As<T, byte>(ref x)] :
+            typeof(T) == typeof(char) ?
+                x => Encoding.UTF8.GetBytes(Unsafe.As<T, char>(ref x).ToString()) :
+                typeof(BitConverter)
+                    .GetMethod(
+                        nameof(BitConverter.GetBytes),
+                        BindingFlags.Public | BindingFlags.Static,
+                        [typeof(T)])!
+                    .CreateDelegate<Func<T, byte[]>>();
+
+        // We're going to assume every char is 1 byte. Technically not correct
+        // since chars are UTF-8, but we're only testing single byte chars 
+        // with this function. I could revert to Marshal.SizeOf here as that
+        // returns 1 for char, but that's not really "correct" as it's intended
+        // for interop.
+        var size = typeof(T) != typeof(char) ? Unsafe.SizeOf<T>() : 1;
+
+        var expectedBytes = expectedUnmanagedValues.SelectMany(getBytes).ToArray();
+        var bytes = new byte[sharedReaderOffset].Concat(expectedBytes).ToArray();
+
+        using var stream = new MemoryStream(bytes);
+        using var binaryReader = new BinaryReader(stream);
+        using var sharedReader = new SharedReader(sharedReaderOffset, binaryReader, new object());
+
+        // Act
+        var actualUnmanagedValues = new T[expectedUnmanagedValues.Length];
```

Comment:
Nitpicking here, but I believe that this step belongs to the `Arrange` section of the unit test. My personal take on this is that the `Act` section should only contain invocation of the method being tested. I see why we include the `for` loop in `Act`, but I think the array initialisation code can be moved to `Arrange`.

Reply:
And done.